### PR TITLE
[Crypto] Switch to libsecp256k1 signature verification and update the lib

### PR DESCRIPTION
**Pull request by Warrows**
Here is the long overdue update for PIVX to let go of OpenSSL in its consensus code.
The rationale behind it is to avoid depending on an external and changing library where our consensus code is affected. This is security and consensus critical. It also brings some performances improvements.
EDIT: Actually, we won't be absolutely free of OpenSSL dependencies because of the zerocoin library. But this is an important first step in this direction and zerocoin lib will follow.

The update is made of 6 commits: 
1. The first 3 are libsecp256k1 update. It's now on the latest version available on https://github.com/bitcoin-core/secp256k1. None of the changes are really mine. By the way, the lib is now handled as a git subtree.
2. 6839f3b Code changes to allow compatibility and actually use this library. That's the bulk of what you'll need to review.
3. 65e009a New auto generated script tests.
4. 21234db is an update after rebasing on top of zPOS.

Here are some highlights for commit 6839f3b and the changes I introduce:
-Removed src/arith_uint256.cpp since its a duplicate implementation of src/uint256.cpp.
-Changed src/bip38.cpp to properly call secp256k1.
-Removed src/eccryptoverify.cpp and src/ecwrapper.cpp since they were used o call OpenSSL crypto.
-Change unsigned char[32] to a struct ChainCode.
-Numerous changes to src/key.cpp and src/pubkey.cpp to comply with libsecp256k1, including:
    -Bringing in some code about keys import/export
    -Change calls from OpenSSL crypto to libsecp256k1
    -Introduce some named values instead of magic numbers
-Move keystore functions implementations from .h to .cpp

Possible testing includes pretty much everything requiring signing or verifying signatures.

This PR is unrelated to #447 and contrary to what I first stated there, PIVX is currently not independent from OpenSSL in its consensus code.
After this PR, the libzerocoin code should be reviewed and if possible uncorrelated from OpenSSL as well.

Performances:
I tried to do an IBD on testnet comparison between this PR and master. But the result ended up being surprising since both were at around 1h10. Actually a bit more for this PR.
However, full IBD also depends on other factors such as download speed. We also have almost empty blocks for which the verification time is not determinant in the perfs. 
The point here is to check if signature validation is faster. And it is.
For example, on master, verifying testnet block 349533 took me 457 ms while with this PR it took 171 ms. 
Since mainnet probably holds more transactions (and hence signatures to check) I expect the IBD to be quicker with this pr on it. Using bootstrap might be much quicker too. 

This PR should not generate too much conflicts with #413 or #416 as they stand. Should it arise, I am prepared to rebase this PR or help rebase the others if this one was to be merged first. EDIT: rebase is done.

### CODE REVIEW

**Comment by Warrows** [comment]
UPDATE: Rebased this PR on top of current master. Merge conflicts with zPoS have been resolved.

**Comment by Fuzzbawls** [comment]
been running with this for the past few days and haven't seen any noticeable issues

**Comment by jonspock** [comment]
Apart from libzerocoin, is there a need to retain dependency for OpenSSL in crypter.cpp?

File: src/bip38.cpp [hunk 1]
```diff
@@ -48,9 +48,23 @@ void ComputePassfactor(std::string ownersalt, uint256 prefactor, uint256& passfa
...
 bool ComputePasspoint(uint256 passfactor, CPubKey& passpoint)
 {
+    size_t clen = 65;
+    secp256k1_context *ctx = NULL;
+    secp256k1_pubkey pubkey;
+
     //passpoint is the ec_mult of passfactor on secp256k1
-    int clen = 65;
-    return secp256k1_ec_pubkey_create(UBEGIN(passpoint), &clen, passfactor.begin(), true) != 0;
+    if (!secp256k1_ec_pubkey_create(ctx, &pubkey, passfactor.begin()))
```
[the remark this answers is not in the document]
Comment (Warrows) [reply src/bip38.cpp]: Great catch, I'll fix it.

**Comment by Warrows** [comment]
@jonspock crypter.cpp is a different question. It's about protecting the wallet file, not creating keys or signing transactions and blocks. But yes, removing openSSL there would be nice too and it should simply be a backport of https://github.com/bitcoin/bitcoin/pull/7689. I don't think libzerocoin has any influence there.

**Comment by jonspock** [comment]
Thanks. Unfortunately the update is not as simple as the diff, but thanks for the hint. Yes, I realized it wasn't consensus related but brought it up because you said " Actually, we won't be absolutely free of OpenSSL dependencies because of the zerocoin library. "

**Comment by Warrows** [comment]
Are we waiting on any more reviews on this?

**Pull request merged**